### PR TITLE
Dojot Provider (Keycloak "Plugin")

### PR DESCRIPTION
**Pull request by thiagodpf**
* **Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

feature

* **What is the current behavior?** (You can also link to an open issue here)

Currently, a simpler solution is used for access management.

* **What is the new behavior (if this is a feature change)?**

We intend to replace the current access control solution with Keycloak, for that, one of the fundamental things that Keycloak needs to do is to publish the tanant creation and removal events on Kafka. So an SPI capable of this was implemented.

Keycloak _Dojot Provider_ is capable of publishing tenant creation events on Kafka.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

no

* **Is there any issue related to this PR? (Link to an issue, e.g. #99999)** 

#1945

* **Other information**:

### CODE REVIEW

File: iam/keycloak/kafka-provider/README.md [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+Keycloak modules are built based on the concept of [Service Provider Interface](https://en.wikipedia.org/wiki/Service_provider_interface) (SPI),
+whose java offers [native support](https://docs.oracle.com/javase/tutorial/sound/SPI-intro.html).
+
+Communication between the Keycloak modules is done through _Service Providers_
+(SPI implementations), which in turn trigger typical events for each module. The
+Keycloak _core_ is responsible for correctly loading all modules and registering
+their event listeners.
+
+In order to be able to capture the events triggered by the Keycloak modules, it
+is necessary to implement an SPI and deploy it to Wildfly so that Keycloak knows
```
Comment (mprevide) [inline iam/keycloak/kafka-provider/README.md:26]: a SPI?

File: iam/keycloak/kafka-provider/README.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+how to load it.
+
+In summary, this README is intended to explain how the _Kafka Provider_ is
+compiled and packaged according th the modular structure of Wildfly. After being
```
Comment (mprevide) [inline iam/keycloak/kafka-provider/README.md:30]: Remove `th`?

File: iam/keycloak/kafka-provider/README.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+```bash
+# In the project's root directory, run the following command:
+cp -a ./target/jboss-modules/. $JBOSS_HOME/modules/system/layers/keycloak/
+```
+
+Now we need to [configure Keycloak](https://www.keycloak.org/docs/latest/server_installation/#_config_spi_providers)
+to recognize our _service provider_ as one of its modules, that is, to
+[load it](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ServiceLoader.html)
+as an SPI.
```
Comment (mprevide) [inline iam/keycloak/kafka-provider/README.md:80]: a SPI?

File: iam/keycloak/examples/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      - zookeeper-volume:/var/lib/zookeeper/data
+      - zookeeper-log-volume:/var/lib/zookeeper/log
+      - zookeeper-secrets-volume:/etc/zookeeper/secrets
+    logging:
+      driver: json-file
+      options:
+        max-size: 20m
+        max-file: '5'
+
+  kafdrop:
```
Comment (mprevide) [inline iam/keycloak/examples/docker-compose.yml:121]: Maybe you can create a readme for this example and explain how to check the kafka message in kafdrop and what needs to be done in the keycloak to create realms.

**Comment by mprevide** [comment]
In addition, you could already create github actions for keycloak. :)

File: iam/keycloak/examples/docker-compose.yml (hunk 4 above)
Comment (thiagodpf) [reply iam/keycloak/examples/docker-compose.yml:121]: done!

**Review by rascaraficci — APPROVED** [review]
Very good job!

In a future version, maybe we can add some metadata to the messages sent to Kafka such as timestamp.

File: .github/workflows/iam-keycloak.yml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+    - name: Build and Push
+      uses: elgohr/Publish-Docker-Github-Action@2.16
+      with:
+        name: ${{ steps.get_owner.outputs.owner }}/keycloak
+        username: ${{ secrets.DOCKER_USERNAME }}
+        password: ${{ secrets.DOCKER_PASSWORD }}
+        workdir: iam/keycloak
+        # Label with the built image's commit
+        buildoptions: " --label 'commit_sha'='${{ github.sha }}' "
```
Comment (mprevide) [inline .github/workflows/iam-keycloak.yml:36]: change to `buildoptions: ' --label commit_sha_dojot=${{ github.sha }} ' `

File: .github/workflows/iam-keycloak.yml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  create:
+  push:
+  pull_request:
+    paths:
+      - 'iam/keycloak/**'
+
+jobs:
+  docker-hub:
+    # Path to help identify each job in the status checks
+    name: (keycloak) Build and Publish image to DockerHub
```
[the remark this answers is not in the document]
Comment (mprevide) [reply .github/workflows/iam-keycloak.yml:13]: So, now this patch is no longer needed .. We can remove all...
Github has changed the way it displays status checks and it is now possible to see the name of the yaml `yaml-name/job`.
image

File: .github/workflows/iam-keycloak.yml (hunk 5 above)
Comment (mprevide) [reply .github/workflows/iam-keycloak.yml:36]: That was not working ok.

File: iam/keycloak/kafka-provider/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (thiagodpf) [reply iam/keycloak/kafka-provider/README.md]: In this case, as the acronym is pronounced starting with the sound of a vowel, it is correct to use the article `an`.

File: iam/examples/keycloak_kafka/docker-compose.yml [hunk 7]
```diff
@@ -14,6 +14,8 @@ services:
       KAFKA_SERVERS: kafka:9092
       KAFKA_CLIENT_ID: keycloak
       KAFKA_TOPIC: dojot-management.dojot.tenancy
+      VALID_REALM_NAME_REGEX: "^[a-zA-Z0-9]{1,30}$$"
+      CUSTOM_REALM_REP_FILE: "/opt/dojot/customRealmRepresentation.json"
```
[the remark this answers is not in the document]
Comment (thiagodpf) [reply iam/examples/keycloak_kafka/docker-compose.yml]: done!

File: iam/keycloak/README.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+# Keycloak for dojot
+
+The [Keycloak](https://www.keycloak.org/) is used by the dojot platform as an
+_Identity and Access Manager_ ([IAM](https://en.wikipedia.org/wiki/Identity_management)).
+
+
+In addition to generating access tokens for the platform's APIs, the Keycloak
+has also been customized to run a specific [provider](https://www.keycloak.org/docs/latest/server_development/index.html#_providers)
```
Comment (mprevide) [inline iam/keycloak/README.md:8]: Maybe change to `https://www.keycloak.org/docs/12.0/server_development/index.html` or to quote the version used in the moment.

Comment (thiagodpf) [reply iam/keycloak/README.md]: Good idea!

File: iam/keycloak/dojot-provider/README.md [hunk 9]
```diff
@@ -194,22 +234,38 @@ $JBOSS_HOME/bin/standalone.sh --server-config=standalone-ha.xml
...
 of the keycloak.
 The file can be included in the keycloak container `VOLUME ["/opt/dojot/"]`.
 Check the [example](../../examples/keycloak_kafka/) directory for more details.
 
 
+## SMTP server configuration
+
+| Environment variable | Reference values | Description |
+| -------------------- | ---------------- | ----------- |
+| DOJOT_SMTP_HOST      | `smtp.gmail.com` | GMail SMTP server host example. |
```
Comment (mprevide) [inline iam/keycloak/dojot-provider/README.md:257]: If the DOJOT_SMPT_ * are empty, will an error occur?

Comment (thiagodpf) [reply iam/keycloak/dojot-provider/README.md:257]: In the `DojotProviderFactoryImpl.java` there is a method called `initSMTPServerConfig()`, If the `DOJOT_SMTP_HOST` variable is not defined or is empty, the SMTP settings will not be passed to the Keycloak.

File: iam/examples/keycloak_kafka/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+      - kafka
+      - postgres
+    restart: always
+    environment:
+      JAVA_OPTS_APPEND: "-Dkeycloak.profile.feature.impersonation=disabled -Dkeycloak.profile.feature.upload_scripts=enabled"
+      KEYCLOAK_USER: admin
+      KEYCLOAK_PASSWORD: admin
+      DOJOT_ROOT_URL: "http://localhost:8080"
+      DOJOT_CUSTOM_REALM_REP_FILE: "/opt/dojot/customRealmRepresentation.json"
+      # DOJOT_ADMIN_PASSWORD: aA1!bB2@cC3#
```
Comment (mprevide) [inline iam/examples/keycloak_kafka/docker-compose.yml:16]: if `DOJOT_ADMIN_PASSWORD` has not been defined, what is the behavior? I think that ideal is cant get up.

Comment (thiagodpf) [reply iam/examples/keycloak_kafka/docker-compose.yml]: Done. This attribute has become mandatory for Keycloak to work!

When trying to create a new Realm, this password will be confronted with the password policies defined in the realm customization json file, if the password is not strong, a message will be displayed on the screen.
In this case, the platform administrator will have to restart the keycloak and set a password that satisfies the password policies.

**Pull request merged**